scalardb/src/scalardb/transfer_append_2pc.clj

```diff
@@ -0,0 +1,263 @@
+         (.put tx (prepare-put i 1 balance)))
+       (.commit tx)))
+   [n]))
+
+(defn- get-balance
+  [^Result result]
+  (-> result (.getValue BALANCE) .get .get))
+
+(defn- get-age
+  [^Result result]
+  (-> result (.getValue AGE) .get .get))
+
+(defn- get-balances
+  [results]
+  (mapv #(get-balance (first %)) results))
+
+(defn- get-ages
+  [results]
+  (mapv #(get-age (first %)) results))
+
+(defn- get-nums
+  [results]
+  (mapv count results))
+
+(defn- calc-new-balance
+  [^Result r amount]
+  (-> r get-balance (+ amount)))
+
+(defn- calc-new-age
+  [^Result r]
+  (-> r get-age inc))
+
+(defn- tx-transfer
+  [tx1 tx2  {:keys [from to amount]}]
+  (try
+    (let [^Result from-result (scan-for-latest tx1 (prepare-scan-for-latest from))
+          ^Result to-result (scan-for-latest tx2 (prepare-scan-for-latest to))]
+      (->> (prepare-put from
+                        (calc-new-age from-result)
+                        (calc-new-balance from-result (- amount)))
+           (.put tx1))
+      (->> (prepare-put to
+                        (calc-new-age to-result)
+                        (calc-new-balance to-result amount))
+           (.put tx2)))
+    (.prepare tx1)
+    (.prepare tx2)
+    (.validate tx1)
+    (.validate tx2)
+    (.commit tx1)
+    (.commit tx2)
+    (catch UnknownTransactionStatusException e
+      (throw e))
+    (catch Exception e
+      (.rollback tx1)
+      (.rollback tx2)
+      (throw e))))
+
+(defn- scan-records
+  [tx id]
+  (try
+    (.scan tx (prepare-scan id))
+    (catch CrudException _ nil)))
+
+(defn- scan-all-records-with-retry
+  [test n]
+  (scalar/check-transaction-connection! test)
+  (scalar/with-retry scalar/prepare-transaction-service! test
+    (let [tx (scalar/start-transaction test)
+          results (map #(scan-records tx %) (range n))]
+      (if (some nil? results) nil results))))
+
+(defrecord TransferClient [initialized? n initial-balance]
+  client/Client
+  (open! [_ _ _]
+    (TransferClient. initialized? n initial-balance))
+
+  (setup! [_ test]
+    (locking initialized?
+      (when (compare-and-set! initialized? false true)
+        (scalar/setup-transaction-tables test [{:keyspace KEYSPACE
+                                                :table TABLE
+                                                :schema SCHEMA}])
+        (scalar/prepare-2pc-service! test)
+        (scalar/prepare-transaction-service! test)
+        (populate-accounts test n initial-balance))))
+
+  (invoke! [_ test op]
+    (case (:f op)
+      :transfer (let [tx1 (scalar/start-2pc test)
+                      tx2 (scalar/join-2pc test (.getId tx1))]
+                  (try
+                    (tx-transfer tx1 tx2 (:value op))
+                    (assoc op :type :ok)
+                    (catch UnknownTransactionStatusException _
+                      (swap! (:unknown-tx test) conj (.getId tx1))
+                      (assoc op :type :fail, :error {:unknown-tx-status (.getId tx1)}))
```

Comment:
@yito88 I just realized this type is `:fail`. Should we change it to `:info`?

If yes, we should also change the same things in `transfer_append.clj`, `transfer.clj`, and `transfer_2pc.clj`.

Reply:
Yes!
It should work because our transfer checker doesn't check `:info`  for now.
However, it should be `:info` since the result isn't checked if it has succeeded or not.
Thanks!

Reply:
Thank you for the confirmation! I will do that in another PR. Thanks.